source/extensions/http/stateful_session/header/header.h

```diff
@@ -0,0 +1,77 @@
+#pragma once
+
+#include <memory>
+#include <tuple>
+
+#include "envoy/extensions/http/stateful_session/header/v3/header.pb.h"
+#include "envoy/http/stateful_session.h"
+
+#include "source/common/common/base64.h"
+#include "source/common/http/headers.h"
+#include "source/common/http/utility.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Http {
+namespace StatefulSession {
+namespace Header {
+
+using HeaderBasedSessionStateProto =
+    envoy::extensions::http::stateful_session::header::v3::HeaderBasedSessionState;
+
+class HeaderBasedSessionStateFactory : public Envoy::Http::SessionStateFactory {
+public:
+  class SessionStateImpl : public Envoy::Http::SessionState {
+  public:
+    SessionStateImpl(absl::optional<std::string> address,
+                     const HeaderBasedSessionStateFactory& factory)
+        : upstream_address_(std::move(address)), factory_(factory) {}
+
+    absl::optional<absl::string_view> upstreamAddress() const override { return upstream_address_; }
+    void onUpdate(const Upstream::HostDescription& host,
+                  Envoy::Http::ResponseHeaderMap& headers) override;
+
+  private:
+    absl::optional<std::string> upstream_address_;
+    const HeaderBasedSessionStateFactory& factory_;
+  };
+
+  HeaderBasedSessionStateFactory(const HeaderBasedSessionStateProto& config);
+
+  Envoy::Http::SessionStatePtr create(const Envoy::Http::RequestHeaderMap& headers) const override {
+    if (!requestPathMatch(headers.getPathValue())) {
+      return nullptr;
+    }
+    return std::make_unique<SessionStateImpl>(parseAddress(headers), *this);
+  }
+
+  bool requestPathMatch(absl::string_view request_path) const {
+    ASSERT(path_matcher_ != nullptr);
+    return path_matcher_(request_path);
+  }
+
+private:
+  absl::optional<std::string> parseAddress(const Envoy::Http::RequestHeaderMap& headers) const {
+    auto hdr = headers.get(Envoy::Http::LowerCaseString(name_));
+    if (hdr.empty()) {
+      return absl::nullopt;
+    }
+
+    auto header_value = hdr[0]->value().getStringView();
+    std::string address = Envoy::Base64::decode(header_value);
+    return !address.empty() ? absl::make_optional(std::move(address)) : absl::nullopt;
+  }
+
+  std::string getHeaderName() const { return name_; }
+
+  const std::string name_;
+  const std::string path_;
+
+  std::function<bool(absl::string_view)> path_matcher_;
```

Comment:
Let's not reinvent the patch_matcher in a filter. This should be achieved by route specific configuration or composite filter.

Reply:
Composite filter looks exciting! A minor issue though:

The Stateful Session Filter supports both listener level (see [here](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/stateful_session_filter#one-example)) and route level configuration ([here](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/stateful_session/v3/stateful_session.proto#extensions-filters-http-stateful-session-v3-statefulsessionperroute))
And this PR is only adding another way to store session state and doesn't aim to change how StatefulSession filter is configured.
Do you think your comment still applies?

Reply:
So route level configuration is already there, why do you need a `path` here? You can disable stateful session by that.

Reply:
That's right. For `StatefulSessionPerRoute`, the path is not needed.
But for `StatefulSession` configuration at Listener level, it is.

My understanding is that:
Stateful Session filter _requires_ two configuration options for any _session storage mechanism_ (such as CookieBasedSessionState)
1. Listener level
2. Route level

Since this PR is only adding another session state storage mechanism (i.e. HeaderBasedSessionState), it has to abide by the laws of StatefulSession filter. 

In short - `HeaderBasedSessionState` must support the Listener level config since `StatefulSession` filter's config dictates that.
@wbpcode to correct me if I misunderstood the API.

Reply:
discussed offline and it makes total sense. I will make the change.